File: src/main/java/com/razorpay/sampleapp/PaymentOptions.java
package com.razorpay.sampleapp;

import android.app.Activity;
import android.content.Intent;
import android.graphics.Bitmap;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.webkit.WebView;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.EditText;
import android.widget.FrameLayout;
import android.widget.LinearLayout;
import android.widget.ListView;
import android.widget.TextView;
import android.widget.Toast;

import com.razorpay.PaymentResultListener;
import com.razorpay.Razorpay;
import com.razorpay.RazorpayWebViewClient;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.Map;

public class PaymentOptions extends Activity implements PaymentResultListener {

    private FrameLayout frameLayout;
    private ListView listView;
    ArrayList<String> banksCodesList = new ArrayList<>();
    private ArrayList<String> banksList = new ArrayList<>();
    private ArrayList<String> walletsList = new ArrayList<>();
    private ArrayAdapter banksListAdapter;
    private ArrayAdapter walletsListAdapter;
    private Razorpay razorpay;
    private WebView webview;
    private ViewGroup outerBox;
    private static final String TAG = PaymentOptions.class.getSimpleName();
    private JSONObject payload;

    private class MethodSelected implements View.OnClickListener {
        @Override
        public void onClick(View view) {
            switch (view.getId()) {

                case R.id.card:
                    frameLayout.removeAllViews();
                    LayoutInflater.from(PaymentOptions.this).inflate(R.layout.fragment_payment_method_card, frameLayout, true);
                    findViewById(R.id.submit_card_details).setOnClickListener(new SubmitCardDetails());
                    razorpay.changeApiKey("rzp_test_1DP5mmOlF5G5ag");
                    break;

                case R.id.upi:
                    frameLayout.removeAllViews();
                    LayoutInflater.from(PaymentOptions.this).inflate(R.layout.fragment_payment_method_upi, frameLayout, true);
                    findViewById(R.id.btn_upi_collect_req).setOnClickListener(new SubmitUPICollectRequest());
                    findViewById(R.id.btn_upi_intent_flow).setOnClickListener(new SubmitUPIIntentDetails());
                    break;

                case R.id.netbanking:
                    frameLayout.removeAllViews();
                    LayoutInflater.from(PaymentOptions.this).inflate(R.layout.fragment_method_netbanking_wallet_list, frameLayout, true);
                    listView = (ListView) findViewById(R.id.method_available_options_list);
                    listView.setAdapter(banksListAdapter);
                    razorpay.changeApiKey("rzp_live_ILgsfZCZoFIKMb");
                    listView.setOnItemClickListener(new AdapterView.OnItemClickListener() {
                        public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
                            submitNetbankingDetails(banksCodesList.get(position));
                        }
                    });
                    break;

                case R.id.wallet:
                    frameLayout.removeAllViews();
                    LayoutInflater.from(PaymentOptions.this).inflate(R.layout.fragment_method_netbanking_wallet_list, frameLayout, true);
                    listView = (ListView) findViewById(R.id.method_available_options_list);
                    listView.setAdapter(walletsListAdapter);
                    listView.setOnItemClickListener(new AdapterView.OnItemClickListener() {
                        public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
                            submitWalletDetails(walletsList.get(position));
                        }
                    });
                    break;

            }
        }
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_payment_options);

        findViewById(R.id.card).setOnClickListener(new MethodSelected());
        findViewById(R.id.upi).setOnClickListener(new MethodSelected());
        findViewById(R.id.netbanking).setOnClickListener(new MethodSelected());
        findViewById(R.id.wallet).setOnClickListener(new MethodSelected());

        webview = (WebView) findViewById(R.id.payment_webview);
        frameLayout = (FrameLayout) findViewById(R.id.frame);
        outerBox = (ViewGroup) findViewById(R.id.outerbox);

        LayoutInflater.from(PaymentOptions.this).inflate(R.layout.fragment_payment_method_card,
                frameLayout, true);
        banksListAdapter = new ArrayAdapter<String>(this, R.layout.text_view_list_banks_wallet, banksList);
        walletsListAdapter = new ArrayAdapter<String>(this, R.layout.text_view_list_banks_wallet, walletsList);

        initRazorpay();
        createWebView();
    }

    private void initRazorpay() {
        razorpay = new Razorpay(this);

        razorpay.getPaymentMethods(new Razorpay.PaymentMethodsCallback() {
            @Override
            public void onPaymentMethodsReceived(String result) {

                /**
                 * This returns JSON data
                 * The structure of this data can be seen at the following link:
                 * https://api.razorpay.com/v1/methods?key_id=rzp_test_1DP5mmOlF5G5ag
                 *
                 */
                Log.d("Result", "" + result);
                inflateLists(result);
            }

            @Override
            public void onError(String error) {

            }
        });
    }

    private void createWebView() {
        /**
         * You need to pass the webview to Razorpay
         */
        razorpay.setWebView(webview);

        /**
         * Override the RazorpayWebViewClient for your custom hooks
         */
        razorpay.setWebviewClient(new RazorpayWebViewClient(razorpay) {

            @Override
            public void onPageStarted(WebView view, String url, Bitmap favicon) {
                super.onPageStarted(view, url, favicon);
                Log.d(TAG, "Custom client onPageStarted");
            }

            @Override
            public void onPageFinished(WebView view, String url) {
                super.onPageFinished(view, url);
                Log.d(TAG, "Custom client onPageFinished");
            }
        });
    }

    private void inflateLists(String result) {
        try {
            JSONObject paymentMethods = new JSONObject(result);
            JSONObject banksListJSON = paymentMethods.getJSONObject("netbanking");
            JSONObject walletListJSON = paymentMethods.getJSONObject("wallet");

            Iterator<String> itr1 = banksListJSON.keys();
            while (itr1.hasNext()) {
                String key = itr1.next();
                banksCodesList.add(key);
                try {
                    banksList.add(banksListJSON.getString(key));
                } catch (JSONException e) {
                    Log.d("Reading Banks List", "" + e.getMessage());
                }
            }

            Iterator<String> itr2 = walletListJSON.keys();
            while (itr2.hasNext()) {
                String key = itr2.next();
                try {
                    if (walletListJSON.getBoolean(key)) {
                        walletsList.add(key);
                    }
                } catch (JSONException e) {
                    Log.d("Reading Wallets List", "" + e.getMessage());
                }
            }

            if (banksListAdapter != null) {
                banksListAdapter.notifyDataSetChanged();
            }

            if (walletsListAdapter != null) {
                walletsListAdapter.notifyDataSetChanged();
            }

        } catch (Exception e) {
            Log.e("Parsing Result", "" + e.getMessage());
        }
    }

    public class SubmitCardDetails implements View.OnClickListener {
        @Override
        public void onClick(View view) {

            EditText nameET = (EditText) findViewById(R.id.name);
            String name = nameET.getText().toString();

            EditText cardNumberET = (EditText) findViewById(R.id.cardNumber);
            String cardNumber = cardNumberET.getText().toString();

            EditText dateET = (EditText) findViewById(R.id.expiry);
            String date = dateET.getText().toString();
            int index = date.indexOf('/');
            String month = date.substring(0, index);
            String year = date.substring(index + 1);

            EditText cvvET = (EditText) findViewById(R.id.cvv);
            String cvv = cvvET.getText().toString();

            try {
                payload = new JSONObject("{currency: 'INR'}");
                payload.put("amount", "100");
                payload.put("contact", "9999999999");
                payload.put("email", "customer@name.com");
            } catch (Exception e) {
                e.printStackTrace();
            }

            try {
                payload.put("method", "card");
                payload.put("card[name]", name);
                payload.put("card[number]", cardNumber);
                payload.put("card[expiry_month]", month);
                payload.put("card[expiry_year]", year);
                payload.put("card[cvv]", cvv);
                sendRequest();
            } catch (Exception e) {
                e.printStackTrace();
            }
        }

    }

    public class SubmitUPICollectRequest  implements View.OnClickListener {
        @Override
        public void onClick(View view) {
            EditText vpaET = (EditText) findViewById(R.id.vpa);
            String vpa = vpaET.getText().toString();

            try {
                payload = new JSONObject("{currency: 'INR'}");
                payload.put("amount", "100");
                payload.put("contact", "9999999999");
                payload.put("email", "customer@name.com");
            } catch (Exception e) {
                e.printStackTrace();
            }


            try {
                payload.put("method", "upi");
                payload.put("vpa", vpa);
                sendRequest();
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
    }

    public class SubmitUPIIntentDetails implements View.OnClickListener {
        @Override
        public void onClick(View view) {
            EditText vpaET = (EditText) findViewById(R.id.vpa);
            String vpa = vpaET.getText().toString();

            try {
                payload = new JSONObject("{currency: 'INR'}");
                payload.put("amount", "100");
                payload.put("contact", "9999999999");
                payload.put("email", "customer@name.com");
                //payload.put("upi_app_package_name", "in.org.npci.upiapp");
                payload.put("display_logo", true);
            } catch (Exception e) {
                e.printStackTrace();
            }

            try {
                JSONArray jArray = new JSONArray();
                jArray.put("in.org.npci.upiapp");
                jArray.put("com.snapwork.hdfc");
                payload.put("method", "upi");
                payload.put("_[flow]", "intent");
                payload.put("preferred_apps_order", jArray);
                payload.put("other_apps_order", jArray);
                sendRequest();
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
    }

    public void submitNetbankingDetails(String bankName) {

        try {
            payload = new JSONObject("{currency: 'INR'}");
            payload.put("amount", "100");
            payload.put("contact", "9999999999");
            payload.put("email", "customer@name.com");
        } catch (Exception e) {
            e.printStackTrace();
        }


        try {
            payload.put("method", "netbanking");
            payload.put("bank", bankName);
            sendRequest();
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public void submitWalletDetails(String walletName) {


        try {
            payload = new JSONObject("{currency: 'INR'}");
            payload.put("amount", "100");
            payload.put("contact", "9999999999");
            payload.put("email", "customer@name.com");
        } catch (Exception e) {
            e.printStackTrace();
        }

        try {
            payload.put("method", "wallet");
            payload.put("wallet", walletName);
        } catch (Exception e) {
            e.printStackTrace();
        }
        sendRequest();
    }

    private void sendRequest() {
        razorpay.validateFields(payload, new Razorpay.ValidationListener() {
            @Override
            public void onValidationSuccess() {
                try {
                    webview.setVisibility(View.VISIBLE);
                    outerBox.setVisibility(View.GONE);
                    razorpay.submit(payload, PaymentOptions.this);
                } catch (Exception e) {
                    Log.e("com.example", "Exception: ", e);
                }
            }

            @Override
            public void onValidationError(Map<String, String> error) {
                Log.d("com.example", "Validation failed: " + error.get("field") + " " + error.get("description"));
                Toast.makeText(PaymentOptions.this, "Validation: " + error.get("field") + " " + error.get("description"), Toast.LENGTH_SHORT).show();
            }
        });
    }

    @Override
    public void onBackPressed() {
        razorpay.onBackPressed();
        super.onBackPressed();
        webview.setVisibility(View.GONE);
        outerBox.setVisibility(View.VISIBLE);
    }

    /* callback for permission requested from android */
    @Override
    public void onRequestPermissionsResult(int requestCode, String[] permissions, int[] grantResults) {
        if (razorpay != null) {
            razorpay.onRequestPermissionsResult(requestCode, permissions, grantResults);
        }
    }

    /**
     * Is called if the payment was successful
     * You can now destroy the webview
     */
    @Override
    public void onPaymentSuccess(String razorpayPaymentId) {
        webview.setVisibility(View.GONE);
        outerBox.setVisibility(View.VISIBLE);
        Toast.makeText(PaymentOptions.this, "Payment Successful: " + razorpayPaymentId, Toast.LENGTH_SHORT).show();
    }

    /**
     * Is called if the payment failed
     * possible values for code in this sdk are:
     * 2: in case of network error
     * 4: response parsing error
     * 5: This will contain meaningful message and can be shown to user
     * Format: {"error": {"description": "Expiry year should be greater than current year"}}
     */
    @Override
    public void onPaymentError(int errorCode, String errorDescription) {
        webview.setVisibility(View.GONE);
        outerBox.setVisibility(View.VISIBLE);
        Toast.makeText(PaymentOptions.this, "Error " + Integer.toString(errorCode) + ": " + errorDescription, Toast.LENGTH_SHORT).show();
        Log.d("com.example", "onError: " + Integer.toString(errorCode) + ": " + errorDescription);
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data){
        super.onActivityResult(requestCode, resultCode, data);
        razorpay.onActivityResult(requestCode,resultCode,data);
    }

}

File: src/main/java/com/razorpay/sampleapp/kotlin/PaymentOptions.kt
package com.razorpay.sampleapp.kotlin

import android.app.Activity
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.webkit.WebView
import android.widget.*
import android.widget.AdapterView.OnItemClickListener
import com.razorpay.*
import com.razorpay.sampleapp.R
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject

class PaymentOptions : Activity(), PaymentResultListener {

    var webView: WebView? = null
    private var frameLayout: FrameLayout? = null
    var outerBox: ViewGroup? = null
    var listView: ListView? = null
    var bankCodesList = ArrayList<String>()
    var banksList = ArrayList<String>()
    var walletsList = ArrayList<String>()
    var banksListAdapter: ArrayAdapter<String>? = null
    var walletsListAdapter: ArrayAdapter<String>? = null
    var razorpay: Razorpay? = null
    var payload = JSONObject("{currency:'INR'}")
    val context = this


    val TAG: String = PaymentOptions::class.toString()


    private class MethodSelected : View.OnClickListener {
        override fun onClick(p0: View?) {
            Log.d("ID","CLICK")
            when (p0?.id) {
                R.id.card -> {
                    PaymentOptions().frameLayout?.removeAllViews()
                    LayoutInflater.from(PaymentOptions().context).inflate(R.layout.fragment_payment_method_card, PaymentOptions().frameLayout, true)
                    PaymentOptions().context.findViewById<Button>(R.id.submit_card_details).setOnClickListener { SubmitCardDetails() }
                }
                R.id.upi -> {
                    PaymentOptions().frameLayout?.removeAllViews()
                    LayoutInflater.from(PaymentOptions().context).inflate(R.layout.fragment_payment_method_upi,PaymentOptions().frameLayout,true)
                    PaymentOptions().context.findViewById<Button>(R.id.btn_upi_collect_req).setOnClickListener { SubmitUPICollectRequest() }
                    PaymentOptions().context.findViewById<Button>(R.id.btn_upi_intent_flow).setOnClickListener { SubmitUPIIntentDetails() }

                }
                R.id.netbanking -> {
                    PaymentOptions().frameLayout?.removeAllViews()
                    LayoutInflater.from(PaymentOptions().context).inflate(R.layout.fragment_method_netbanking_wallet_list, PaymentOptions().frameLayout, true)
                    PaymentOptions().listView = PaymentOptions().context.findViewById<View>(R.id.method_available_options_list) as ListView
                    PaymentOptions().listView?.adapter = PaymentOptions().banksListAdapter
                    PaymentOptions().razorpay?.changeApiKey("rzp_live_ILgsfZCZoFIKMb")
                    PaymentOptions().listView?.setOnItemClickListener { parent, view, position, id ->
                        run {
                            PaymentOptions().submitNetbankingDetails(PaymentOptions().bankCodesList[position])
                        }
                    }
                }
                R.id.wallet -> {
                    PaymentOptions().frameLayout?.removeAllViews()
                    LayoutInflater.from(PaymentOptions().context).inflate(R.layout.fragment_method_netbanking_wallet_list,PaymentOptions().frameLayout,true)
                    PaymentOptions().listView = PaymentOptions().context.findViewById(R.id.method_available_options_list)
                    PaymentOptions().listView?.adapter = PaymentOptions().walletsListAdapter
                    PaymentOptions().listView?.setOnItemClickListener { parent, view, position, id ->
                        run{
                            PaymentOptions().submitWalletDetails(PaymentOptions().walletsList[position])
                        }
                    }
                }
            }
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_payment_options)

        findViewById<TextView>(R.id.card).setOnClickListener { MethodSelected() }
        findViewById<TextView>(R.id.upi).setOnClickListener { MethodSelected() }
        findViewById<TextView>(R.id.netbanking).setOnClickListener {
            frameLayout?.removeAllViews()
            LayoutInflater.from(this).inflate(R.layout.fragment_method_netbanking_wallet_list, PaymentOptions().frameLayout, true)

            listView = findViewById(R.id.method_available_options_list)
            listView?.adapter = banksListAdapter

            razorpay?.changeApiKey("rzp_live_ILgsfZCZoFIKMb")
            listView?.setOnItemClickListener { parent, view, position, id ->
                run {
                    PaymentOptions().submitNetbankingDetails(PaymentOptions().bankCodesList[position])
                }
            }
        }
        findViewById<TextView>(R.id.wallet).setOnClickListener { MethodSelected() }

        webView = findViewById(R.id.payment_webview)
        frameLayout = findViewById(R.id.frame)
        outerBox = findViewById(R.id.outerbox)

        LayoutInflater.from(this).inflate(R.layout.fragment_payment_method_card, frameLayout, true)
        banksListAdapter = ArrayAdapter(this, R.layout.text_view_list_banks_wallet, banksList)
        walletsListAdapter = ArrayAdapter(this, R.layout.text_view_list_banks_wallet, walletsList)

        initRazorpay()
        createWebView()

    }


    private fun createWebView() {
        razorpay?.setWebView(webView)
    }

    private fun initRazorpay() {
        razorpay = Razorpay(this)

        razorpay?.getPaymentMethods(object : PaymentMethodsCallback {
            override fun onPaymentMethodsReceived(result: String?) {
                /**
                 * This returns JSON data
                 * The structure of this data can be seen at the following link:
                 * https://api.razorpay.com/v1/methods?key_id=rzp_test_1DP5mmOlF5G5ag
                 *
                 */
                Log.d("Result", "" + result)
                inflateLists(result)
            }

            override fun onError(error: String?) {
                if (error != null) {
                    Log.e("Get Payment error", error)
                }
            }
        })

        razorpay?.isValidVpa("stambatgr5@okhdfcbank", object : ValidateVpaCallback {
            override fun onResponse(p0: JSONObject?) {
            }

            override fun onFailure() {
                Toast.makeText(this@PaymentOptions, "Error validating VPA", Toast.LENGTH_LONG).show()
            }
        })
    }

    private fun inflateLists(result: String?) {
        try {
            val paymentMethods = JSONObject(result)
            val banksListJSON = paymentMethods.getJSONObject("netbanking")
            val walletListJSON = paymentMethods.getJSONObject("wallet")

            val itr1: Iterator<String> = banksListJSON.keys()
            while (itr1.hasNext()) {
                val key = itr1.next()
                bankCodesList.add(key)
                try {
                    banksList.add(banksListJSON.getString(key))
                } catch (jsonException: JSONException) {
                    Log.e("Reading Banks List", jsonException.localizedMessage)
                }
            }

            val itr2: Iterator<String> = walletListJSON.keys()
            while (itr2.hasNext()) {
                val key = itr2.next()
                try {
                    if (walletListJSON.getBoolean(key)) {
                        walletsList.add(key)
                    }
                } catch (jsonException: JSONException) {
                    Log.e("Reading Wallets List", jsonException.localizedMessage)
                }
            }

            banksListAdapter?.notifyDataSetChanged()
            walletsListAdapter?.notifyDataSetChanged()

        } catch (e: Exception) {
            Log.e("Parsing Result", e.localizedMessage)
        }
    }

    public class SubmitCardDetails : View.OnClickListener {

        override fun onClick(view: View?) {
            val context = PaymentOptions().context

            val etName = context.findViewById<EditText>(R.id.name)
            val name = etName.text.toString()

            val etCardNumber = context.findViewById<EditText>(R.id.cardNumber)
            val cardNumber = etCardNumber.text.toString()

            val etDate = context.findViewById<EditText>(R.id.expiry)
            val date = etDate.text.toString()

            val index = date.indexOf('/')
            val month = date.substring(0, index)
            val year = date.substring(index + 1)

            val etCvv = context.findViewById<EditText>(R.id.cvv)
            val cvv = etCvv.text.toString()
            val payload = PaymentOptions().payload
            try {
                payload.put("amount", "100")
                payload.put("contact", "9999999999")
                payload.put("email", "customer@name.com")
            } catch (e: Exception) {
                e.printStackTrace()
            }

            try {
                payload.put("method", "card")
                payload.put("card[name]", name)
                payload.put("card[number]", cardNumber)
                payload.put("card[expiry_month]", month)
                payload.put("card[expiry_year]", year)
                payload.put("card[cvv]", cvv)
                PaymentOptions().sendRequest()
            } catch (e: Exception) {
                e.printStackTrace()
            }

        }


    }

    public class SubmitUPICollectRequest : View.OnClickListener{
        override fun onClick(view: View?) {
            val context = PaymentOptions().context
            val etVpa = context.findViewById<EditText>(R.id.vpa)
            val vpa = etVpa.text.toString()
            var payload = PaymentOptions().payload
            try{
                payload = JSONObject("{currency: 'INR'}")
                payload.put("amount", "100")
                payload.put("contact", "9999999999")
                payload.put("email", "customer@name.com")
            }catch (e: Exception){
                e.printStackTrace()
            }

            try{
                payload.put("method","upi")
                payload.put("vpa",vpa)
                PaymentOptions().sendRequest()
            }catch (e: Exception){
                e.printStackTrace()
            }

        }

    }

    class SubmitUPIIntentDetails : View.OnClickListener {
        override fun onClick(view: View) {
            val context = PaymentOptions().context
            val vpaET = context.findViewById<View>(R.id.vpa) as EditText
            val vpa = vpaET.text.toString()
            val payload = PaymentOptions().payload
            try {
                payload.put("amount", "111")
                payload.put("contact", "9999999999")
                payload.put("email", "customer@name.com")
                //payload.put("upi_app_package_name", "com.google.android.apps.nbu.paisa.user");
                payload.put("display_logo", true)
            } catch (e: Exception) {
                e.printStackTrace()
            }
            try {
                val jArray = JSONArray()
                jArray.put("in.org.npci.upiapp")
                jArray.put("com.snapwork.hdfc")
                //payload.put("key_id","rzp_test_kEVtCVFWAjUQPG");
                payload.put("method", "upi")
                payload.put("_[flow]", "intent")
                payload.put("preferred_apps_order", jArray)
                payload.put("other_apps_order", jArray)
                PaymentOptions().sendRequest()
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }

    fun submitNetbankingDetails(bankName: String?) {
        try {
            payload = JSONObject("{currency: 'INR'}")
            payload.put("amount", "100")
            payload.put("contact", "9999999999")
            payload.put("email", "customer@name.com")
        } catch (e: Exception) {
            e.printStackTrace()
        }
        try {
            payload.put("method", "netbanking")
            payload.put("bank", bankName)
            sendRequest()
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    fun submitWalletDetails(walletName: String?) {
        try {
            payload = JSONObject("{currency: 'INR'}")
            payload.put("amount", "100")
            payload.put("contact", "9999999999")
            payload.put("email", "customer@name.com")
        } catch (e: Exception) {
            e.printStackTrace()
        }
        try {
            payload.put("method", "wallet")
            payload.put("wallet", walletName)
        } catch (e: Exception) {
            e.printStackTrace()
        }
        sendRequest()
    }

    private fun sendRequest() {
        razorpay?.validateFields(payload, object : ValidationListener {
            override fun onValidationError(error: MutableMap<String, String>?) {
                Log.d(TAG, "Validation failed: " + error?.get("field"))
                Toast.makeText(this@PaymentOptions, "Validation: " + error?.get("field"), Toast.LENGTH_LONG).show()
            }

            override fun onValidationSuccess() {
                try {
                    webView?.visibility = View.VISIBLE
                    outerBox?.visibility = View.GONE
                    razorpay?.submit(payload, this@PaymentOptions)
                } catch (e: Exception) {
                    Log.e(TAG, "Excpetion: ", e)
                    e.printStackTrace()
                }
            }

        })
    }

    override fun onBackPressed() {
        razorpay?.onBackPressed()
        super.onBackPressed()
        webView?.visibility = View.GONE
        outerBox?.visibility = View.VISIBLE
    }

    /**
     * Is called if the payment failed
     * possible values for code in this sdk are:
     * 2: in case of network error
     * 4: response parsing error
     * 5: This will contain meaningful message and can be shown to user
     * Format: {"error": {"description": "Expiry year should be greater than current year"}}
     */
    override fun onPaymentError(errorCode: Int, errorDescription: String?) {
        webView?.visibility = View.GONE
        outerBox?.visibility = View.VISIBLE
        Toast.makeText(this@PaymentOptions, "Error $errorCode : $errorDescription",Toast.LENGTH_LONG).show()
        Log.e(TAG,"onError: $errorCode : $errorDescription")
    }

    /**
     * Is called if the payment was successful
     * You can now destroy the webview
     */
    override fun onPaymentSuccess(rzpPaymentId: String?) {
        webView?.visibility = View.GONE
        outerBox?.visibility = View.VISIBLE
        Toast.makeText(this@PaymentOptions, "Payment Successful: $rzpPaymentId",Toast.LENGTH_LONG).show()
    }
}
































































